## 예외

### 1. 예외 블랙홀

```java
try {
    ...
}
catch (SQLException e) {
}
```

```java
} catch (SQLException e) {
    e.printStackTrace();
}
```

위 코드는 예외를 잡고는 아무것도 하지 않는다. 예외 발생을 무시해버리고 정상적인 상황인 것처럼 다음 라인으로 넘어가겠다는 분명한 의도가 있는 게 아니라면 만들어서는 안 되는 코드다. 예외가 발생하면 화면에 출력하는 경우도 예외를 처리한 것은 아니다.

> 모든 예외는 적절하게 복구되거나 작업을 중단시키고 분명하게 통보되어야 한다

### 2.예외의 종류와 특징

![img](https://github.com/dilmah0203/TIL/blob/main/Image/Exception01.png)

- **Error** : 시스템에 비정상적인 상황이 발생했을 경우에 사용된다. 애플리케이션에서는 이런 에러에 대한 처리는 신경쓰지 않아도 된다.

- **Exception과 체크 예외** : java.lang.Exception 클래스는 체크 예외와 언체크 예외로 구분된다. 일반적으로 예외라고 하면 체크 예외를 말한다. 체크 예외가 발생할 수 있는 메소드를 사용할 경우 반드시 예외를 처리하는 코드를 작성해야 한다.

- **RuntimeException과 언체크/런타임 예외** : 런타임 예외는 프로그램의 오류가 있을 때 발생하도록 의도된 것들이다. 대표적으로 오브젝트를 할당하지 않은 래퍼런스 변수를 사용할 때 발생하는 NullPointerException이 있다. 런타임 예외는 예상하지 못했던 예외상황에서 발생하는 것이 아니기 때문에 굳이 catch나 throws를 사용하지 않아도 된다.

### 3.예외처리 방법

1. **예외 복구**

예외상황을 파악하고 문제를 해결해서 정상 상태로 돌려놓는 것이다. 예를들어 사용자가 요청한 파일을 읽으려고 할 때 해당 파일이 없다거나 다른 문제가 있어 IOException이 발생한 경우, 사용자에게 상황을 알려주고 다른 파일을 이용하도록 하여 예외상황을 해결할 수 있다. 예외로 인해 기본 작업 흐름이 불가능하면 다른 작업 흐름으로 유도하는 것이다. 단, IOException 에러 메세지가 사용자에게 그냥 던져지는 것은 예외 복구라고 볼 수 없다.

2. **예외처리 회피**

예외처리를 자신이 담당하지 않고 자신을 호출한 쪽으로 던져버리는 것이다. throws 문으로 선언해서 예외가 발생하면 알아서 던져지게 하거나 catch 문으로 예외를 잡은 후 로그를 남기고 다시 예외를 던지는 것이다.

```java
public void add() throws SQLException {
    try {
    }
    catch (SQLException e) {
        //로그 출력
        throw e;
    }
}
```

3. **예외 전환**

예외 회피와 비슷하게 예외를 복구해서 정상적인 상태로는 만들 수 없기 때문에 예외를 메소드 밖으로 던지는 것이다. 하지만 예외 회피와는 달리, 발생한 예외를 그대로 넘기는 게 아니라 적절한 예외로 전환해서 던진다는 특징이 있다. 

예외 전환은 두 가지 목적으로 사용된다. 
- **내부에서 발생한 예외를 그대로 던지는 것이 그 예외상황에 대한 적절한 의미를 부여해주지 못하는 경우에 의미를 분명하게 해주기 위해**

예를 들어 새로운 사용자를 등록하려고 시도했을 때 아이디가 같은 사용자가 있어서 DB 에러가 발생하면 JDBC API는 SQLException을 발생시킨다. 이 경우 DAO 메소드가 SQLException을 그대로 밖으로 던져버리면, 서비스 계층 등에서는 왜 SQLException이 발생했는지 쉽게 알 방법이 없다. 이럴 땐 DAO에서 SQLException의 정보를 해석해서 DuplicateUserIdException 같은 예외로 바꾸어서 던져주는 것이 좋다. 의미가 분명한 예외가 던져지면 서비스 계층 오브젝트에는 적절한 복구 작업을 시도할 수 있다.

```java
public void add(User user) throws DuplicateUserIdException, SQLException {
    try{
        //JDBC를 이용해 user 정보를 DB에 추가하는 코드
    }
    catch (SQLException e) {
        //ErrorCode가 MySQL의 "Duplicate Entry"이면 예외 전환
        if (e.getErrorCode() == MysqlErrorNumbers.ER_DUP_ENTRY)
            throw DuplicateUserIdException();
        else
            throw e;
    }
}
```

보통 전환하는 예외에 원래 발생한 예외를 담아서 **중첩 예외**로 만드는 것이 좋다. 중첩 예외는 getCause() 메소드를 이용해 처음 발생한 예외가 무엇인지 확인할 수 있다.

```java
catch (SQLExcetpion e) {
    throw DuplicateUserIdException(e); //throw DuplicateUserIdException().initCause(e);
}
```


- **예외를 처리하기 쉽고 단순하게 만들기 위해 포장하려는 목적**

중첩 예외를 이용해 새로운 예외를 만들고 원인이 되는 예외를 내부에 담아서 던지는 방식은 같지만, 예외처리를 강제하는 체크예외를 언체크 예외인 런타임 예외로 바꾸는 경우에 사용한다.

